### PR TITLE
Provide support for machete-pre-rebase hook

### PR DESCRIPTION
**Pull request by PawelLipski**
~Still requires better error/timeout handling~ done

### CODE REVIEW

File: scripts/git-hooks/machete-pre-rebase [hunk 1]
```diff
@@ -13,14 +13,51 @@ if [ $# -eq 3 ]; then
...
 
   if [[ $head_pre_release ]]; then
-    if [[ $head_pre_release -ne 1 ]]; then
-      target_version="$base_major.$base_minor.$base_patch-$((${base_pre_release:-0} + 1))"
-    elif [[ $head_patch -ne 0 ]]; then
-      target_version="$base_major.$base_minor.$((base_patch + 1))-1"
-    elif [[ $head_minor -ne 0 ]]; then
-      target_version="$base_major.$((base_minor + 1)).0-1"
+    if [[ $base_pre_release ]]; then
+      # Both head and base versions are pre-release. That's the most unintuitive case.
```
Comment (mkondratek) [inline scripts/git-hooks/machete-pre-rebase:17]: whoooa :books:  great explanations!

Comment (PawelLipski) [reply scripts/git-hooks/machete-pre-rebase:17]: :bow:

File: scripts/git-hooks/machete-pre-rebase [hunk 2]
```diff
@@ -40,5 +77,6 @@ if [ $# -eq 3 ]; then
 elif [ $# -eq 1 ]; then
   target_version=$1
 
-  sed -i "/^\s*PLUGIN_VERSION = '.*'\$/  s/'.*'/'$target_version'/" version.gradle
+  # Using -i='' and not just -i to retain compatibility with both Linux (GNU) sed and OS X (BSD) sed.
+  sed -i='' "/^\s*PLUGIN_VERSION = '.*'\$/  s/'.*'/'$target_version'/" version.gradle
```
Comment (PawelLipski) [inline scripts/git-hooks/machete-pre-rebase:81]: `-i.bak` + `rm -f version.gradle.bak`

File: backendImpl/src/main/java/com/virtuslab/gitmachete/backend/impl/hooks/PreRebaseHookExecutor.java [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+    //   Before Git invokes a hook, it changes its working directory to either $GIT_DIR in a bare repository
+    //   or the root of the working tree in a non-bare repository.
+    //   An exception are hooks triggered during a push (...) which are always executed in $GIT_DIR.
+    // We obviously assume a non-bare repository here, and machete-pre-rebase isn't related to push.
+    pb.directory(mainDirectory);
+
+    Process process;
+    try {
+      process = pb.start();
+      int timeoutSeconds = 10;
```
Comment (micpiotrowski) [inline backendImpl/src/main/java/com/virtuslab/gitmachete/backend/impl/hooks/PreRebaseHookExecutor.java:65]: Maybe move this var to class field - it seems to be something more class related than local code related

File: scripts/git-hooks/machete-pre-rebase (hunk 2 above)
Comment (PawelLipski) [reply scripts/git-hooks/machete-pre-rebase]: Done

File: backendImpl/src/main/java/com/virtuslab/gitmachete/backend/impl/hooks/PreRebaseHookExecutor.java (hunk 3 above)
Comment (PawelLipski) [reply backendImpl/src/main/java/com/virtuslab/gitmachete/backend/impl/hooks/PreRebaseHookExecutor.java:65]: Done on another PR for both this executor and the one for `machete-status-branch`

**Pull request merged**